src/uniswap/incentives.ts

```diff
@@ -0,0 +1,446 @@
+    let provider = providers[providerAddress];
+    if (provider === undefined) {
+      provider = providers[providerAddress] = new ProviderLiquidity();
+    }
+
+    const liquidityMs: bigint = liquidity * BigInt(timeRange);
+
+    provider.liquidity = provider.liquidity + liquidityMs;
+    rangeLiquidity = rangeLiquidity + liquidityMs;
+  };
+
+  const { balances, blocks } = balancesStore;
+
+  /*
+   * TODO An algorithm here is inefficient, as we go through all the positions for every minute.
+   * Considering we are moving in time in one direction, it would have been way more efficient, for
+   * example, to sort positions, and look only at a subset that matches the current time.
+   *
+   * Note that we have both time and price constraints.  But there are even more efficient
+   * approaches.
+   *
+   * For now, considering there are no automated tests, I opted for the most dump implementation,
+   * hoping that the runtime would still be bearable.
+   */
+
+  /*
+   * How many digits of precision to use when adjusting liquidity using price and time coefficients.
+   * As all the liquidity computations use BigInt, we need to do calculations using integers.
+   * So we are going to multiply values by `10n ** coefficientsDigits`, before we divide them by the
+   * inverse at the end of the computation.
+   */
+  const coefficientsDigits = 16n;
+
+  const { providerEvents } = balances;
+  for (const providerAddress in providerEvents) {
+    const priceRanges = providerEvents[providerAddress];
+
+    for (const priceRange of priceRanges) {
+      const { min: rangePriceMin, max: rangePriceMax, events } = priceRange;
+
+      const effectivePriceMin = Math.max(priceMin, rangePriceMin);
+      const effectivePriceMax = Math.min(priceMax, rangePriceMax);
+
+      if (effectivePriceMin >= effectivePriceMax) {
+        /* This position is completely outside of the specified ranges. */
+        continue;
+      }
+
+      const priceCoefficient = BigInt(
+        Math.floor(
+          ((effectivePriceMax - effectivePriceMin) / (priceMax - priceMin)) *
+            10 ** Number(coefficientsDigits)
+        )
+      );
+
+      for (const event of events) {
+        const block = blocks[event.block];
+
+        if (block.timestamp.getTime() >= to.getTime()) {
+          continue;
+        }
+
+        let timeCoefficient: bigint;
+        let liquidityAdjustment: bigint;
+
+        switch (event.event) {
+          case "mint":
+            timeCoefficient = BigInt(
+              Math.floor(
+                (differenceInMilliseconds(to, max([block.timestamp, from])) /
+                  timeRange) *
+                  10 ** Number(coefficientsDigits)
+              )
+            );
+            liquidityAdjustment = event.liquidity;
+            break;
+
+          case "burn":
+            timeCoefficient = BigInt(
+              Math.floor(
+                (differenceInMilliseconds(max([to, block.timestamp]), from) /
```

Comment:
I think this should be the same as for mint.

Reply:
Overall I think we could be simpler

f(t, provider) = sum_{events: time(event) < t} (t-time(event) * deltaLiquidity(event)

then we are interested in f(t_end,provider) - f(t_start, provider)

Reply:
Oh of course forgot the binance price window restriction. So we have to iterate over the one minute time intervals but the point is there is no difference between mint/burn except for the sign of the deltaLiquidity